### PR TITLE
Support `thrust::count, device` in CuPy JIT

### PR DESCRIPTION
**Pull request by TsutsuiMasayoshi**
It is a part of PFN summer internship project made by @TsutsuiMasayoshi (1 / n)
This PR is a part of supporting use of thrust inside CuPy JIT.

### CODE REVIEW

File: cupyx/jit/thrust.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+from cupyx.jit import _cuda_types
+
+
+def _wrap_thrust_func(header):
+    def wrapper(func):
+        class FuncWrapper(_internal_types.BuiltinFunc):
+            def call(self, env, *args, **kwargs):
+                env.generated.add_code(f'#include <{header}>')
+                env.generated.add_code('#include <thrust/functional.h>')
+                env.generated.backend = 'nvcc'
```
Comment (emcastillo) [inline cupyx/jit/thrust.py:11]: Could we use nvrtc here and enable jitify?
In theory it should work
(originally asked offline by @leofang)

Comment (asi1024) [reply cupyx/jit/thrust.py:11]: I'm trying jitify to include thrust with the minimal sample code, but the compilation did not succeed. Are there any missing options? cc: @leofang

```py
import cupy

source = """
#include <thrust/pair.h>

extern "C" __global__ void test_kernel(float *arr) {}
"""

kern = cupy.RawKernel(
    source,
    'test_kernel',
    backend='nvrtc',
    jitify=True,
)

x = cupy.arange(12, dtype='f')
```

```
---------------------------------------------------
--- JIT compile log for /tmp/tmpt6o3opqw/ceaed42501056bbc8416fd9e6c561b85c4a84188.cubin.cu ---
---------------------------------------------------
thrust/detail/config/cpp_dialect.h(131): warning #161-D: unrecognized #pragma

thrust/detail/tuple.inl(78): error: namespace "std" has no member "add_const"

thrust/detail/tuple.inl(78): error: expected a ";"

thrust/detail/tuple.inl(84): error: namespace "std" has no member "add_volatile"

thrust/detail/tuple.inl(84): error: expected a ";"

thrust/detail/tuple.inl(90): error: namespace "std" has no member "add_cv"

thrust/detail/tuple.inl(90): error: expected a ";"

6 errors detected in the compilation of "/tmp/tmpt6o3opqw/ceaed42501056bbc8416fd9e6c561b85c4a84188.cubin.cu".
```

Comment (asi1024) [reply cupyx/jit/thrust.py:11]: @leofang If including also `thurst/excecution_policy.h`, so many additional compilation errors are caused :cry:.
It seems to require so many fixes to resolve all of them, so I would like to use `nvcc` in this pull request.

Error messages:

<details>

```
thrust/detail/memory_wrapper.h(29): warning: memory: [jitify] File not found
thrust/detail/allocator/allocator_traits.inl(27): warning: new: [jitify] File not found
---------------------------------------------------
--- JIT compile log for /tmp/tmpcn68i_yl/131725912a8520f354cf7501bf117ccfca7cd64d.cubin.cu ---
---------------------------------------------------
thrust/detail/config/cpp_dialect.h(131): warning #161-D: unrecognized #pragma

thrust/detail/tuple.inl(78): error: namespace "std" has no member "add_const"

thrust/detail/tuple.inl(78): error: expected a ";"

thrust/detail/tuple.inl(84): error: namespace "std" has no member "add_volatile"

thrust/detail/tuple.inl(84): error: expected a ";"

thrust/detail/tuple.inl(90): error: namespace "std" has no member "add_cv"

thrust/detail/tuple.inl(90): error: expected a ";"

thrust/detail/alignment.h(153): warning #1866-D: attribute does not apply to any entity

thrust/detail/allocator/allocator_traits.h(57): error: namespace "std" has no member "declval"

thrust/detail/allocator/allocator_traits.h(57): error: type name is not allowed

thrust/detail/allocator/allocator_traits.h(57): error: expected an expression

limits.h(19): error: "_JITIFY_CHAR_IS_UNSIGNED" has already been declared in the current scope

limits.h(20): error: "CHAR_MIN" has already been declared in the current scope

limits.h(21): error: "CHAR_MAX" has already been declared in the current scope

thrust/detail/allocator/allocator_traits.inl(43): error: namespace "std" has no member "allocator"

thrust/detail/allocator/allocator_traits.inl(43): error: expected a ">"

thrust/detail/allocator/allocator_traits.inl(70): warning #12-D: parsing restarts here after previous syntax error

thrust/detail/allocator/allocator_traits.inl(75): error: identifier "pointer" is undefined

thrust/detail/type_traits/pointer_traits.h(201): error: name followed by "::" must be a class or namespace name
          detected during instantiation of class "thrust::detail::pointer_traits<Ptr> [with Ptr=<error-type>]" 
thrust/detail/allocator/allocator_traits.inl(75): here

thrust/detail/type_traits/pointer_traits.h(202): error: incomplete type is not allowed
          detected during instantiation of class "thrust::detail::pointer_traits<Ptr> [with Ptr=<error-type>]" 
thrust/detail/allocator/allocator_traits.inl(75): here

thrust/detail/type_traits/pointer_traits.h(72): error: name followed by "::" must be a class or namespace name
          detected during:
            instantiation of class "thrust::detail::pointer_difference<Ptr> [with Ptr=<error-type>]" 
(203): here
            instantiation of class "thrust::detail::pointer_traits<Ptr> [with Ptr=<error-type>]" 
thrust/detail/allocator/allocator_traits.inl(75): here

thrust/detail/type_traits/pointer_traits.h(212): error: class "thrust::detail::pointer_traits_detail::pointer_to_param<<error-type>>" has no member "type"
          detected during instantiation of class "thrust::detail::pointer_traits<Ptr> [with Ptr=<error-type>]" 
thrust/detail/allocator/allocator_traits.inl(75): here

thrust/detail/type_traits/pointer_traits.h(221): error: class "thrust::detail::pointer_raw_pointer<<error-type>>" has no member "type"
          detected during instantiation of class "thrust::detail::pointer_traits<Ptr> [with Ptr=<error-type>]" 
thrust/detail/allocator/allocator_traits.inl(75): here

thrust/detail/allocator/allocator_traits.inl(76): error: identifier "const_pointer" is undefined

thrust/detail/allocator/allocator_traits.inl(79): error: namespace "std" has no member "allocator"

thrust/detail/allocator/allocator_traits.inl(79): error: expected a ";"

thrust/detail/allocator/allocator_traits.inl(81): error: namespace "std" has no member "allocator"

thrust/detail/allocator/allocator_traits.inl(81): error: expected a ">"

thrust/detail/allocator/allocator_traits.inl(81): error: expected a ";"

thrust/detail/allocator/allocator_traits.inl(85): error: identifier "pointer" is undefined

thrust/detail/allocator/allocator_traits.inl(85): error: identifier "allocator_type" is undefined

thrust/detail/allocator/allocator_traits.inl(85): error: identifier "a" is undefined

thrust/detail/allocator/allocator_traits.inl(85): error: identifier "size_type" is undefined

thrust/detail/allocator/allocator_traits.inl(85): error: expected a ")"

thrust/detail/allocator/allocator_traits.inl(86): error: expected a ";"

thrust/detail/allocator/allocator_traits.inl(90): error: this pragma must immediately precede a declaration

thrust/detail/allocator/allocator_traits.inl(97): error: this pragma must immediately precede a declaration

thrust/detail/allocator/allocator_traits.inl(104): error: this pragma must immediately precede a declaration

thrust/detail/allocator/allocator_traits.inl(112): error: this pragma must immediately precede a declaration

thrust/detail/allocator/allocator_traits.inl(120): error: this pragma must immediately precede a declaration

thrust/detail/allocator/allocator_traits.inl(126): warning #12-D: parsing restarts here after previous syntax error

thrust/detail/allocator/allocator_traits.inl(133): error: namespace "std" has no member "declval"

thrust/detail/allocator/allocator_traits.inl(133): error: type name is not allowed

thrust/detail/allocator/allocator_traits.inl(133): error: expected an expression

thrust/detail/allocator/allocator_traits.inl(133): error: add_reference is not a template

thrust/detail/allocator/allocator_traits.inl(138): error: allocator_traits is not a template

thrust/detail/allocator/allocator_traits.inl(139): error: allocator_traits is not a template

thrust/detail/allocator/allocator_traits.inl(140): error: allocator_traits is not a template

thrust/detail/allocator/allocator_traits.inl(149): error: enable_if is not a template

thrust/detail/allocator/allocator_traits.inl(151): error: allocator_traits is not a template

thrust/detail/allocator/allocator_traits.inl(151): error: type name is not allowed

thrust/detail/allocator/allocator_traits.inl(153): error: allocator_traits is not a template

thrust/detail/allocator/allocator_traits.inl(153): error: allocator_traits is not a template

thrust/detail/allocator/allocator_traits.inl(160): error: disable_if is not a template

thrust/detail/allocator/allocator_traits.inl(162): error: allocator_traits is not a template

thrust/detail/allocator/allocator_traits.inl(162): error: type name is not allowed

thrust/detail/allocator/allocator_traits.inl(164): error: allocator_traits is not a template

thrust/detail/allocator/allocator_traits.inl(164): error: allocator_traits is not a template

thrust/detail/allocator/allocator_traits.inl(164): error: function template "thrust::allocator_traits_detail::allocate" has already been defined

thrust/detail/allocator/allocator_traits.inl(170): error: namespace "std" has no member "declval"

thrust/detail/allocator/allocator_traits.inl(170): error: type name is not allowed

thrust/detail/allocator/allocator_traits.inl(170): error: expected an expression

thrust/detail/allocator/allocator_traits.inl(170): error: add_reference is not a template

thrust/detail/allocator/allocator_traits.inl(180): error: enable_if is not a template

thrust/detail/allocator/allocator_traits.inl(191): error: disable_if is not a template

thrust/detail/allocator/allocator_traits.inl(194): error: function template "thrust::allocator_traits_detail::construct" has already been defined

thrust/detail/allocator/allocator_traits.inl(200): error: namespace "std" has no member "declval"

thrust/detail/allocator/allocator_traits.inl(200): error: type name is not allowed

thrust/detail/allocator/allocator_traits.inl(200): error: expected an expression

thrust/detail/allocator/allocator_traits.inl(200): error: add_reference is not a template

thrust/detail/allocator/allocator_traits.inl(210): error: enable_if is not a template

thrust/detail/allocator/allocator_traits.inl(221): error: disable_if is not a template

thrust/detail/allocator/allocator_traits.inl(224): error: function template "thrust::allocator_traits_detail::construct(Alloc &, T *, const Arg1 &)" has already been defined

thrust/detail/allocator/allocator_traits.inl(231): error: namespace "std" has no member "declval"

thrust/detail/allocator/allocator_traits.inl(231): error: type name is not allowed

thrust/detail/allocator/allocator_traits.inl(231): error: expected an expression

thrust/detail/allocator/allocator_traits.inl(231): error: add_reference is not a template

thrust/detail/allocator/allocator_traits.inl(241): error: enable_if is not a template

thrust/detail/allocator/allocator_traits.inl(252): error: disable_if is not a template

thrust/detail/allocator/allocator_traits.inl(255): error: function template "thrust::allocator_traits_detail::construct(Alloc &, T *, Args &&...)" has already been defined

thrust/detail/allocator/allocator_traits.inl(262): error: namespace "std" has no member "declval"

thrust/detail/allocator/allocator_traits.inl(262): error: type name is not allowed

thrust/detail/allocator/allocator_traits.inl(262): error: expected an expression

thrust/detail/allocator/allocator_traits.inl(262): error: add_reference is not a template

thrust/detail/allocator/allocator_traits.inl(272): error: enable_if is not a template

thrust/detail/allocator/allocator_traits.inl(283): error: disable_if is not a template

thrust/detail/allocator/allocator_traits.inl(286): error: function template "thrust::allocator_traits_detail::destroy" has already been defined

thrust/detail/allocator/allocator_traits.inl(292): error: namespace "std" has no member "declval"

thrust/detail/allocator/allocator_traits.inl(292): error: type name is not allowed

thrust/detail/allocator/allocator_traits.inl(292): error: expected an expression

thrust/detail/allocator/allocator_traits.inl(292): error: add_reference is not a template

thrust/detail/allocator/allocator_traits.inl(297): error: allocator_traits is not a template

thrust/detail/allocator/allocator_traits.inl(306): error: enable_if is not a template

thrust/detail/allocator/allocator_traits.inl(308): error: allocator_traits is not a template

thrust/detail/allocator/allocator_traits.inl(308): error: type name is not allowed

thrust/detail/allocator/allocator_traits.inl(317): error: disable_if is not a template

thrust/detail/allocator/allocator_traits.inl(319): error: allocator_traits is not a template

thrust/detail/allocator/allocator_traits.inl(319): error: type name is not allowed

thrust/detail/allocator/allocator_traits.inl(321): error: function template "thrust::allocator_traits_detail::max_size" has already been defined

thrust/detail/allocator/allocator_traits.inl(329): error: enable_if is not a template

thrust/detail/allocator/allocator_traits.inl(330): error: identifier "has_member_system" is undefined

thrust/detail/allocator/allocator_traits.inl(330): error: type name is not allowed

thrust/detail/allocator/allocator_traits.inl(352): warning #12-D: parsing restarts here after previous syntax error

thrust/detail/allocator/allocator_traits.inl(352): error: expected a ";"

thrust/detail/allocator/allocator_traits.inl(357): error: allocator_traits is not a template

Error limit reached.
100 errors detected in the compilation of "/tmp/tmpcn68i_yl/131725912a8520f354cf7501bf117ccfca7cd64d.cubin.cu".
Compilation terminated.

---------------------------------------------------

```

Comment (leofang) [reply cupyx/jit/thrust.py:11]: Hi @asi1024, I tried your above reproducer (with `thrust/pair.h` included) but it worked fine. Am I missing something?

Comment (asi1024) [reply cupyx/jit/thrust.py:11]: @leofang Sorry, kernel call `kern((1,), (1,), (x,))` was missing... Fixed the reproducer!

Comment (leofang) [reply cupyx/jit/thrust.py:11]: @asi1024 Thanks, but I forgot to mention I did already add `kern.compile()` to force compilation, and neither that nor your updated reproducer would give me a compilation error... What's the CUDA version you're using?

Comment (asi1024) [reply cupyx/jit/thrust.py:11]: @leofang Here is my result of `cupy.show_config()`. Jitify does not support CUDA 11.x series?

```
OS                           : Linux-4.15.0-175-generic-x86_64-with-glibc2.27
Python Version               : 3.9.7
CuPy Version                 : 12.0.0a1
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 1.23.3
SciPy Version                : 1.9.1
Cython Build Version         : 0.29.28
Cython Runtime Version       : 0.29.28
CUDA Root                    : /usr/local/cuda-11.6.0
nvcc PATH                    : /usr/local/cuda-11.6.0/bin/nvcc
CUDA Build Version           : 11060
CUDA Driver Version          : 11060
CUDA Runtime Version         : 11060
cuBLAS Version               : (available)
cuFFT Version                : 10600
cuRAND Version               : 10209
cuSOLVER Version             : (11, 3, 2)
cuSPARSE Version             : (available)
NVRTC Version                : (11, 6)
Thrust Version               : 101500
CUB Build Version            : 101500
Jitify Build Version         : 4a37de0
cuDNN Build Version          : None
cuDNN Version                : None
NCCL Build Version           : None
NCCL Runtime Version         : None
cuTENSOR Version             : None
cuSPARSELt Build Version     : None
```

Comment (leofang) [reply cupyx/jit/thrust.py:11]: Thanks. I will check. I suspect it's due to different Thrust versions. I was using CUDA 11.4.

**Review by emcastillo — APPROVED** [review]
LGTM!

**Comment by emcastillo** [comment]
Fails for rocm :(

```
FAILED cupyx_tests/jit_tests/test_thrust.py::TestThrust::test_count_iterator[F]



16:35:35 E               In file included from /opt/rocm/include/thrust/system/hip/detail/count.h:34:
16:35:35 E               /opt/rocm/include/thrust/system/hip/detail/reduce.h:153:11: error: no matching constructor for initialization of 'thrust::hip_rocprim::transform_input_iterator_t<long, CArrayIterator<int, int>, thrust::detail::equal_to_value<unsigned int>>'
16:35:35 E                         first,
```

**Comment by asi1024** [comment]
Fixed to xfail rocm test!

**Comment by asi1024** [comment]
@emcastillo Passed rocm test!

**Review by emcastillo — APPROVED** [review]

**Pull request merged**